File: clients/windows/Mailcal.Tests/EventEditorStateTests.cs
// The event editor's decisions, tested without a dialog: the two payload shapes (a zoned wall-clock
// create, an event-zone edit), the all-day inclusive↔exclusive conversion that bites both ways, the
// fields frozen on edit, validity, and the reminder bucketing. Pure values, the wall clocks are plain
// numbers held apart from any zone, so the assertions hold whatever zone the test machine is in. The
// Windows twin of Apple's EventEditorStateTests and Android's EventEditorTest.
using System;
using Allodia.Mailcal.Calendar;
using uniffi.mailcal_bindings;
using Xunit;

namespace Allodia.Mailcal.Tests;

public class EventEditorStateTests
{
    // A fixed "now", built from components so the wall clock is stable on any machine.
    private static readonly DateTime Now = new(2026, 8, 1, 9, 15, 0, DateTimeKind.Unspecified);

    private static EventDetail Detail(
        bool allDay,
        string timezone,
        string start,
        string end,
        bool isRecurring = false,
        int? reminderMinutes = null,
        EventRecurrence? recurrence = null,
        EventAttendee[]? attendees = null) =>
        new(
            "acct",
            "/cal/e.ics",
            "work",
            "Standup",
            allDay,
            timezone,
            start,
            end,
            "Room 2",
            "bring the roadmap",
            reminderMinutes,
            recurrence,
            null,
            isRecurring,
            true,
            null,
            attendees ?? []);

    [Fact]
    public void Created_timed_event_is_a_wall_clock_in_the_device_zone()
    {
        var editor = EventEditorState.Create(
            new CalendarChoice("acct", "work", "Work"), "Europe/Amsterdam", Now);
        editor.Title = "Lunch";
        editor.Location = "Room 6"; // a create can set a location now (engine PR)
        var args = editor.CreateArgs();
        Assert.Equal("Lunch", args.Title);
        Assert.Equal("2026-08-01T10:00:00", args.Start); // next whole hour after 09:15
        Assert.Equal("2026-08-01T11:00:00", args.End);
        Assert.Equal("Europe/Amsterdam", args.Timezone);
        Assert.Equal("acct", args.Account);
        Assert.Equal("work", args.Calendar);
        Assert.False(args.AllDay);
        Assert.Null(args.Notes);
        Assert.Equal("Room 6", args.Location);
    }

    [Fact]
    public void Created_event_with_no_location_sends_none()
    {
        // Empty stays absent, the core turns a null into no LOCATION line at all.
        var editor = EventEditorState.Create(
            new CalendarChoice("acct", "work", "Work"), "Europe/Amsterdam", Now);
        editor.Title = "Lunch";
        Assert.Null(editor.CreateArgs().Location);
    }

    [Fact]
    public void Created_all_day_event_sends_an_exclusive_end_and_no_zone()
    {
        var editor = EventEditorState.Create(
            new CalendarChoice("acct", "work", "Work"), "Europe/Amsterdam", Now);
        editor.Title = "Holiday";
        editor.AllDay = true; // one day: start and end both fall on 2026-08-01
        var args = editor.CreateArgs();
        Assert.True(args.AllDay);
        Assert.Equal("2026-08-01", args.Start);
        Assert.Equal("2026-08-02", args.End); // exclusive
        Assert.Null(args.Timezone);
    }

    [Fact]
    public void Editing_prefills_the_own_wall_clock_and_updates_it()
    {
        var editor = EventEditorState.Edit(
            Detail(false, "Europe/Amsterdam", "2026-01-05T09:30:00", "2026-01-05T10:00:00"), "Work");
        Assert.True(editor.IsEditing);
        Assert.Equal("Standup", editor.Title);
        Assert.Equal("Room 2", editor.Location);

        editor.Title = "Standup (kort)";
        var args = editor.UpdateArgs(thisOccurrenceOnly: false);
        Assert.Equal("acct", args.Account);
        Assert.Equal("/cal/e.ics", args.Key);
        Assert.Equal("Standup (kort)", args.Title);
        Assert.Equal("2026-01-05T09:30:00", args.Start);
        Assert.Equal("2026-01-05T10:00:00", args.End);
        Assert.Equal("Room 2", args.Location);
        Assert.Null(args.Occurrence); // v1 edits the whole series
    }

    [Fact]
    public void Editing_an_all_day_event_shows_the_inclusive_day_and_saves_the_exclusive_one()
    {
        // The detail's end is exclusive (04-02 for a one-day event on the 1st). The editor must show
        // the 1st and save the 2nd again, an off-by-one here grows a one-day event to two.
        var editor = EventEditorState.Edit(Detail(true, "", "2026-04-01", "2026-04-02"), "Work");
        Assert.True(editor.AllDay);
        var args = editor.UpdateArgs(thisOccurrenceOnly: false);
        Assert.Equal("2026-04-01", args.Start);
        Assert.Equal("2026-04-02", args.End);
    }

    [Fact]
    public void All_day_and_calendar_are_frozen_on_edit_but_free_on_create()
    {
        Assert.True(EventEditorState.Create(null, "Europe/Amsterdam", Now).CanEditForm);
        Assert.False(
            EventEditorState.Edit(
                Detail(false, "Europe/Amsterdam", "2026-01-05T09:30:00", "2026-01-05T10:00:00"), "Work")
                .CanEditForm);
    }

    [Fact]
    public void An_editor_is_invalid_without_a_title_or_a_positive_interval()
    {
        var editor = EventEditorState.Create(
            new CalendarChoice("acct", "work", "Work"), "Europe/Amsterdam", Now);
        Assert.False(editor.IsValid); // blank title
        editor.Title = "X";
        Assert.True(editor.IsValid);
        editor.End = editor.Start;
        Assert.False(editor.IsValid); // end must be after start
    }

    [Fact]
    public void Reminders_bucket_into_the_coarsest_exact_unit()
    {
        Assert.IsType<ReminderBucket.None>(CalendarEventSummary.ReminderBucketOf(null));
        Assert.IsType<ReminderBucket.AtStart>(CalendarEventSummary.ReminderBucketOf(0));
        Assert.Equal(new ReminderBucket.Minutes(15), CalendarEventSummary.ReminderBucketOf(15));
        Assert.Equal(new ReminderBucket.Hours(2), CalendarEventSummary.ReminderBucketOf(120));
        Assert.Equal(new ReminderBucket.Days(1), CalendarEventSummary.ReminderBucketOf(1440));
    }
}

File: clients/apple/Packages/MailcalKit/Tests/MailcalUITests/EventEditorStateTests.swift
// The event editor's decisions, tested without a view: the two payload shapes (a zoned wall-clock
// create, an event-zone edit), the all-day inclusive↔exclusive conversion that bites both ways, the
// fields frozen on edit, and validity. Pure values, the wall clocks round-trip through the device
// calendar, so the assertions hold whatever zone the test machine is in.

import Foundation
import MailcalBindings
import Testing

@testable import MailcalUI

@Suite struct EventEditorStateTests {
    // A fixed "now", built from components so the wall clock is stable on any machine.
    private var now: Date {
        Calendar.current.date(from: DateComponents(year: 2026, month: 8, day: 1, hour: 9, minute: 15))!
    }

    private func detail(
        allDay: Bool,
        timezone: String,
        start: String,
        end: String,
        isRecurring: Bool = false,
        reminderMinutes: Int32? = nil,
        recurrence: EventRecurrence? = nil,
        attendees: [EventAttendee] = []
    ) -> EventDetail {
        EventDetail(
            account: "acct",
            key: "/cal/e.ics",
            calendar: "work",
            title: "Standup",
            allDay: allDay,
            timezone: timezone,
            start: start,
            end: end,
            location: "Room 2",
            notes: "bring the roadmap",
            reminderMinutes: reminderMinutes,
            recurrence: recurrence,
            repeatSummary: nil,
            repeatDraft: nil,
            isRecurring: isRecurring,
            canWrite: true,
            occurrenceStart: "",
            attendees: attendees
        )
    }

    @Test func createdTimedEventIsAWallClockInTheDeviceZone() {
        var editor = EventEditorState.create(
            default: CalendarChoice(account: "acct", id: "work", name: "Work"),
            zone: "Europe/Amsterdam",
            now: now
        )
        editor.title = "Lunch"
        editor.location = "Room 6" // a create can set a location now (engine PR)
        let args = editor.createArgs()
        #expect(args.title == "Lunch")
        #expect(args.start == "2026-08-01T10:00:00") // next whole hour after 09:15
        #expect(args.end == "2026-08-01T11:00:00")
        #expect(args.timezone == "Europe/Amsterdam")
        #expect(args.account == "acct")
        #expect(args.calendar == "work")
        #expect(args.allDay == false)
        #expect(args.notes == nil)
        #expect(args.location == "Room 6")
    }

    @Test func createdEventWithNoLocationSendsNone() {
        // Empty stays absent, the core turns a nil into no LOCATION line at all.
        var editor = EventEditorState.create(
            default: CalendarChoice(account: "acct", id: "work", name: "Work"),
            zone: "Europe/Amsterdam",
            now: now
        )
        editor.title = "Lunch"
        #expect(editor.createArgs().location == nil)
    }

    @Test func createdAllDayEventSendsAnExclusiveEndAndNoZone() {
        var editor = EventEditorState.create(
            default: CalendarChoice(account: "acct", id: "work", name: "Work"),
            zone: "Europe/Amsterdam",
            now: now
        )
        editor.title = "Holiday"
        editor.allDay = true // one day: start and end both fall on 2026-08-01
        let args = editor.createArgs()
        #expect(args.allDay == true)
        #expect(args.start == "2026-08-01")
        #expect(args.end == "2026-08-02") // exclusive
        #expect(args.timezone == nil)
    }

    @Test func editingPrefillsTheOwnWallClockAndUpdatesIt() {
        var editor = EventEditorState.edit(
            detail(allDay: false, timezone: "Europe/Amsterdam", start: "2026-01-05T09:30:00", end: "2026-01-05T10:00:00"),
            calendarName: "Work"
        )
        #expect(editor.isEditing)
        #expect(editor.title == "Standup")
        #expect(editor.location == "Room 2")

        editor.title = "Standup (kort)"
        let args = editor.updateArgs(thisOccurrenceOnly: false)
        #expect(args.account == "acct")
        #expect(args.key == "/cal/e.ics")
        #expect(args.title == "Standup (kort)")
        #expect(args.start == "2026-01-05T09:30:00")
        #expect(args.end == "2026-01-05T10:00:00")
        #expect(args.location == "Room 2")
        #expect(args.occurrence == nil) // v1 edits the whole series
    }

    @Test func editingAnAllDayEventShowsTheInclusiveDayAndSavesTheExclusiveOne() {
        // The detail's end is exclusive (04-02 for a one-day event on the 1st). The editor must show
        // the 1st and save the 2nd again, an off-by-one here grows a one-day event to two.
        let editor = EventEditorState.edit(
            detail(allDay: true, timezone: "", start: "2026-04-01", end: "2026-04-02"),
            calendarName: "Work"
        )
        #expect(editor.allDay)
        let args = editor.updateArgs(thisOccurrenceOnly: false)
        #expect(args.start == "2026-04-01")
        #expect(args.end == "2026-04-02")
    }

    @Test func allDayAndCalendarAreFrozenOnEditButFreeOnCreate() {
        #expect(EventEditorState.create(default: nil, zone: "Europe/Amsterdam", now: now).canEditForm)
        #expect(
            EventEditorState.edit(
                detail(allDay: false, timezone: "Europe/Amsterdam", start: "2026-01-05T09:30:00", end: "2026-01-05T10:00:00"),
                calendarName: "Work"
            ).canEditForm == false
        )
    }

    @Test func anEditorIsInvalidWithoutATitleOrAPositiveInterval() {
        var editor = EventEditorState.create(
            default: CalendarChoice(account: "acct", id: "work", name: "Work"),
            zone: "Europe/Amsterdam",
            now: now
        )
        #expect(editor.isValid == false) // blank title
        editor.title = "X"
        #expect(editor.isValid)
        editor.end = editor.start
        #expect(editor.isValid == false) // end must be after start
    }

    @Test func aRefAsksAboutTheSeriesExactlyWhenItNamesAnOccurrence() {
        // The guard on the delete path: the detail is reached from a grid block, a month chip or
        // an agenda row, and only the first two name the day the user was looking at. An agenda
        // row is the series, so a delete from there is a series delete and asks nothing.
        let fromTheGrid = EventRefID(
            account: "acct", key: "/cal/e.ics", occurrence: "2026-08-04T09:00:00"
        )
        let fromTheAgenda = EventRefID(account: "acct", key: "/cal/e.ics", occurrence: "")
        #expect(fromTheGrid.asksAboutTheSeries)
        #expect(fromTheAgenda.asksAboutTheSeries == false)
    }

    @Test func remindersBucketIntoTheCoarsestExactUnit() {
        #expect(reminderBucket(nil) == .none)
        #expect(reminderBucket(0) == .atStart)
        #expect(reminderBucket(15) == .minutes(15))
        #expect(reminderBucket(120) == .hours(2))
        #expect(reminderBucket(1440) == .days(1))
    }
}
